### DISCUSSION TITLE
How to perform `EXCLUSIVE` transactions?

### DISCUSSION THREAD

**Discussion by benallfree** [pocketbase/pocketbase#752] [Q&A]
What is the recommended way to start an `EXCLUSIVE` transaction in SQLite?

I've traced through the code and it seems that [`dao.runInTransaction`](https://pkg.go.dev/github.com/pocketbase/pocketbase@v0.7.9/daos#Dao.RunInTransaction) boils down to [`dbx.Begin`](https://pkg.go.dev/github.com/pocketbase/dbx#DB.Begin) without any args, which I think in SQLite will default to `DEFERRED` according to https://www.sqlite.org/lang_transaction.html.

**Comment by ganigeorgiev** [comment]
`dbx` internally uses `database/sql` to start the transaction and unfortunately the currently available config struct [`database/sql.TxOptions`](https://pkg.go.dev/database/sql#TxOptions) has only `IsolationLevel` option but this is not equivalent to the SQLite transaction types. There is an open issue in the std library to add support for this in https://github.com/golang/go/issues/19981.

I'm not sure if there is other solution than executing raw statements.

To be honest I can't think of an app scenario where I would want to prevent db reads. Could you elaborate on your use case?

**Reply by ganigeorgiev** [reply]
Actually on second read of the sqlite docs, the `EXCLUSIVE` and `IMMEDIATE` are the same in WAL mode so it seems that it will not prevent db reads, but will return an error if another write transaction has already started. Is that what you want?

**Reply by benallfree** [reply]
I'm trying to understand how SQLite handles race conditions. If I do a series of reads in a transaction, the docs say they are consistent even if outside writes occur. But then if I do conditional writes in the transaction based upon the read results - which have now changed externally - that seems like a problem.

**Reply by ganigeorgiev** [reply]
Yes, multiple "read" only transactions are ok, but once you have "read+write", the transaction will get upgraded to a "write" or will return a SQLITE_BUSY (if there is another write currently running).

**Reply by ganigeorgiev** [reply]
It depends a lot on what you are trying to do, but long running transactions in SQLite are usually a bottleneck and you should try to keep them short and light as much as possible to avoid hitting the SQLITE_BUSY error.
